keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1089,23 +1089,78 @@ repeated evaluations if they are shared across validations.
 
 #### Secondary Authz
 
-We have general agreement to include this as a feature, but need to provide
-a concrete design.
-
-kube-apiserver authorizer checks (aka Secondary-authz checks) have been proposed
-as a way of doing things like:
+TODO: background and motivation
 
-- Validate that only a user with a specific permission can set a particular
+- Validate that only a user with a specific permission can set a particular field.
+- Validate that only a controller responsible for a finalizer can remove it from the finalizers
   field.
-- Validate that only a controller responsible for a finalizer can remove it from
-  the finalizers field.
 
-This could be supported by matching criteria, or via CEL expression access, or both.
- 
-Use cases:
+To depend on an authz decision, validation expressions can reference the identifier `authorizer`,
+which will be bound at evaluation time to an Authorizer object supporting receiver-style function
+overloads:
+
+| Symbol      | Type                                                 | Description |
+|-------------|------------------------------------------------------|-------------|
+| path        | Authorizer.(string) -> PathCheck                     | todo        |
+| check       | PathCheck.(string) -> Decision                       | todo        |
+| resource    | Authorizer.(string, string, string) -> ResourceCheck | todo        |
+| subresource | ResourceCheck.(string) -> ResourceCheck              | todo        |
+| namespace   | ResourceCheck.(string) -> ResourceCheck              | todo        |
+| name        | ResourceCheck.(string) -> ResourceCheck              | todo        |
+| check       | ResourceCheck.(string) -> Decision                   | todo        |
+| allowed     | Decision.() -> bool                                  | todo        |
+| denied      | Decision.() -> bool                                  | todo        |
+
+Example expressions using `authorizer`:
+
+- `authorizer.resource('signers', 'certificates.k8s.io', '*').name(oldObject.spec.signerName).check('approve').allowed()`
+- `authorizer.path('/metrics').check('get').denied()`
+
+Note that this API:
+
+- Produces errors at compilation time when parameters are missing or improperly combined
+  (e.g. subresource with non-resource path, missing path or resource).
+- Is open to the addition of future knobs (e.g. impersonation).
+
+Other considerations:
+
+- Does there need to be a special limit on the number of authz checks generated during expression
+  evaluation, or is it sufficient to assign an appropriate fixed cost?
```

Comment:
Really good question. Including a fixed cost makes seems valuable, and should be easy to do. If we set that cost high enough, a dedicated limit might not be needed, but I don't have a good sense of how many authz checks would be "too many".

Reply:
It seems like a variable number of authorization checks might be a common use case? Consider the policy: "You must be authorized to read every secret your pod mounts" - This would require making an authorization for each secret referenced.

Reply:
That's a great example.